lib/OpenQA/Worker/Job.pm

```diff
@@ -0,0 +1,1263 @@
+    my $chunk_size = $self->worker->settings->global_settings->{UPLOAD_CHUNK_SIZE} // 1000000;
+    my $ua         = $self->client->ua;
+    my $error;
+
+    log_info("Uploading $filename using multiple chunks", channels => ['worker'], default => 1);
+
+    $ua->upload->once(
+        'upload_chunk.prepare' => sub {
+            my ($self, $pieces) = @_;
+            log_info("$filename: " . $pieces->size() . " chunks",   channels => ['worker'], default => 1);
+            log_info("$filename: chunks of $chunk_size bytes each", channels => ['worker'], default => 1);
+        });
+    my $t_start;
+    $ua->upload->on('upload_chunk.start' => sub { $t_start = time() });
+    $ua->upload->on(
+        'upload_chunk.finish' => sub {
+            my ($self, $piece) = @_;
+            my $spent  = (time() - $t_start) || 1;
+            my $kbytes = ($piece->end - $piece->start) / 1024;
+            my $speed  = sprintf("%.3f", $kbytes / $spent);
+            log_info(
+                "$filename: Processing chunk " . $piece->index() . "/" . $piece->total . " avg speed ~${speed}KB/s",
+                channels => ['worker'],
+                default  => 1
+            );
+        });
+
+    $ua->upload->on(
+        'upload_chunk.response' => sub {
+            my ($self, $res) = @_;
+            if ($res->res->is_server_error) {
+                log_error($res->res->json->{error}, channels => ['autoinst', 'worker'], default => 1)
+                  if $res->res->json && $res->res->json->{error};
+                my $msg = "Failed uploading chunk";
+                log_error($msg, channels => ['autoinst', 'worker'], default => 1);
+            }
+            _log_upload_error($filename, $res);
+        });
+    $ua->upload->on(
+        'upload_chunk.fail' => sub {
+            my ($self, $res, $chunk) = @_;
+            log_error(
+                "Upload failed for chunk " . $chunk->index,
+                channels => ['autoinst', 'worker'],
+                default  => 1
+            );
+            sleep 5;    # do not choke webui
+        });
+
+    $ua->upload->once(
+        'upload_chunk.error' => sub {
+            $error = pop();
+            log_error(
+                "Upload failed, and all retry attempts have been exhausted",
+                channels => ['autoinst', 'worker'],
+                default  => 1
+            );
+        });
+
+    local $@;
+    eval {
+        $ua->upload->asset(
+            $job_id => {
+                file       => $file,
+                name       => $filename,
+                asset      => $upload_parameter->{asset},
+                chunk_size => $chunk_size,
+            });
+    };
+    log_error($@, channels => ['autoinst', 'worker'], default => 1) if $@;
+
+    $ua->upload->unsubscribe($_)
+      for qw(upload_chunk.request_err upload_chunk.error upload_chunk.fail),
+      qw( upload_chunk.response upload_chunk.start upload_chunk.finish upload_chunk.prepare);
+
+    return 0 if $@ || $error;
```

Comment:
I would save `$@` to a variable directly after the `eval`. Any other command called in between can change it and that can be hard to debug (e.g. the error message can be lost).
I had such a case in my previous project.
I always recommend to do `my $err = $@`  directly after the eval, if you are going to use it more than once.

Reply:
The mix of `$@`, `$_` and the `local $@;` before the `eval` scares me. I'm not 100% sure what this code is doing. Can you suggest more concrete how I should change this (without breaking existing behavior)? Note that I have not written this code. I've just moved it to a different location.

Reply:
```perl
    local $@;
    eval {
        $ua->upload->asset(
        # ...
    };
    my $err = $@;
    log_error($err, channels => ['autoinst', 'worker'], default => 1) if $err;
# ...
    return 0 if $err || $error;
```
I haven't checked where `$error` comes from, though.

Reply:
@perlpunk Why are you using `local $@` there?